hackathon_site/dashboard/frontend/src/slices/users/userSlice.test.js

```diff
@@ -74,3 +74,145 @@ describe("userData Reducers", () => {
         });
     });
 });
+
+describe("logIn Thunk and Reducer", () => {
+    const body = { email: "foo@bar.com", password: "foobar123" };
+    let store;
+
+    beforeEach(() => {
+        store = mockStore(mockState);
+    });
+
+    describe("Reducers", () => {
+        test("Pending", () => {
+            expect(reducer(initialState, logIn.pending()).login.isLoading).toBe(true);
+        });
+
+        test("Fulfilled", () => {
+            expect(reducer(initialState, logIn.fulfilled())).toEqual(
+                expect.objectContaining({
+                    isAuthenticated: true,
+                    login: { isLoading: false, failure: null },
+                })
+            );
+        });
+
+        test("Rejected", () => {
+            const failureResponse = { status: 999, message: "Some message" };
+            const action = logIn.rejected(
+                "Rejected", // createAsyncThunk will always have {message: "Rejected"} for rejections
+                "some-id",
+                body,
+                failureResponse // This becomes action.payload
+            );
+            expect(reducer(initialState, action)).toEqual(
+                expect.objectContaining({
+                    isAuthenticated: false,
+                    login: { isLoading: false, failure: failureResponse },
+                })
+            );
+        });
+    });
+
+    test("Successful login", async () => {
```

Comment:
Pasting in my comments from our slack convo for posterity:

I'm not sure about the caveats of mocking dispatch instead of going for a full solution like `redux-mock-store`. I agree that for this case where all I'm doing is getting actions, it will work. Everywhere I've seen online uses `redux-mock-store` for testing dispatches anyway, and we already have it installed for wrapping components in a store for tests.

There's some unnecessary testing by checking all dispatch calls / store actions, since we know that the pending action will always be dispatched. Really all that I care about is that push was dispatched (need to test the action payload for that as well), and that fulfilled was dispatched with the appropriate payload.

Good call on not wanting to rely on the library's implementation of requestId, I've fixed the tests to get around that. I'm conflicted about how I hard-coded the expected {type: ..., payload: ...} action structure as opposed to using the action creator itself, but I'm pretty confident that redux-toolkit will never change that structure (which would break our reducers if it did, anyway)